Selection rects for Live Text for images with vertical writing mode are sometimes incorrect

rr-codes opened the pull request:
#### 4cf2bc3696e0fc7a634c1f73a2ba9d42bfa0d298
<pre>
Selection rects for Live Text for images with vertical writing mode are sometimes incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=255625">https://bugs.webkit.org/show_bug.cgi?id=255625</a>
rdar://106205724

Reviewed by Wenson Hsieh and Megan Gardner.

For Live Text, always use `UseIndivudalQuads` so that non-rectilinear selections always work.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::adjustLineHeightOfSelectionGeometries):
(WebCore::RenderObject::collectSelectionGeometriesInternal):
(WebCore::coalesceSelectionGeometryWithAdjacentQuadsIfPossible):
(WebCore::RenderObject::collectSelectionGeometries):

Ensure that adjacent rects are coalesced if they overlap.

* Source/WebKit/UIProcess/ios/WKTextSelectionRect.mm:
(-[WKTextSelectionRectCustomHandleInfo initWithFloatQuad:isHorizontal:]):
(-[WKTextSelectionRectCustomHandleInfo bottomLeft]):
(-[WKTextSelectionRectCustomHandleInfo topLeft]):
(-[WKTextSelectionRectCustomHandleInfo bottomRight]):
(-[WKTextSelectionRectCustomHandleInfo topRight]):
(-[WKTextSelectionRect _customHandleInfo]):
(-[WKTextSelectionRect isVertical]):
(-[WKTextSelectionRectCustomHandleInfo initWithFloatQuad:]): Deleted.

Work around UIKit issues with the selection handle by changing the quad points.

Canonical link: <a href="https://commits.webkit.org/263159@main">https://commits.webkit.org/263159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73ae89f1fd4387529f9e4bca8d82c49a590ee15b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3818 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3906 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4014 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5250 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4077 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3792 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3984 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3911 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3660 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3864 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4064 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3438 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5084 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1568 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3413 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/4520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3387 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3472 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4849 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3869 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3394 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3413 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/926 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3435 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3671 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/platform/graphics/FloatQuad.cpp:
```diff
@@ -109,6 +109,21 @@ bool FloatQuad::containsQuad(const FloatQuad& other) const
+    for (const auto& p : { other.p1(), other.p2(), other.p3(), other.p4() }) {
+        if (containsPoint(p))
+            return true;
+    }
```
Hm....I don't think this takes care of some corner cases (such as this case):

```
░░░░░░█████████████░░░░
░░░░░░█░░░░░░░░░░░█░░░░
░░░░░░█░░░░░░░░░░░█░░░░
░│││││█│││││││││││█│││░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│░░░░█░░░░░░░░░░░█░░│░
░│││││█│││││││││││█│││░
░░░░░░█░░░░░░░░░░░█░░░░
░░░░░░█░░░░░░░░░░░█░░░░
░░░░░░█░░░░░░░░░░░█░░░░
░░░░░░█████████████░░░░
```

(It might also be good to write a quick unit test case for this in `FloatQuadTests.cpp`)

aprotyas commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2548,13 +2548,36 @@ static bool coalesceSelectionGeometryWithAdjacentQuadsIfPossible(SelectionGeomet
+    auto areOverlapping = [](const FloatQuad& first, const FloatQuad& second) {
+        bool intersects = false;
+
+        for (const auto& p : { first.p1(), first.p2(), first.p3(), first.p4() }) {
```
Oh interesting, what is the brace constructing? I assumed `FloatQuad` but that type doesn't have `begin()`/`end()` methods?

rr-codes replied:
It creates an `std::initializer_list` and then iterates over the values inside it

whsieh replied:
I am _pretty sure_ it's just a `std::initializer_list<FloatPoint>`.

Edit: yeah, that seems right. Tiny test program:

```
#include <WebCore/FloatPoint.h>

using namespace WebCore;

int main(int argumentCount, const char* arguments[])
{
    const auto points = { FloatPoint { }, FloatPoint { }, FloatPoint { } };
    points.thisMethodDoesNotExist();

    return 0;
}
```

...with compiler error:

```
main.cpp:10:12: error: no member named 'thisMethodDoesNotExist' in 'std::initializer_list<WebCore::FloatPoint>'
    points.thisMethodDoesNotExist();
    ~~~~~~ ^
1 error generated.
```

rr-codes replied:
it's actually a list of `FloatPoint` I believe, not `FloatQuad`

aprotyas replied:
Thanks for the explanation!

aprotyas commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2548,13 +2548,36 @@ static bool coalesceSelectionGeometryWithAdjacentQuadsIfPossible(SelectionGeomet
+        auto isTrailingSideOverlapping = (first.p1() - first.p2()).diagonalLength() + (second.p1() - second.p2()).diagonalLength() > (first.p1() - second.p2()).diagonalLength();
+        auto isLeadingSideOverlapping = (first.p3() - first.p4()).diagonalLength() + (second.p3() - second.p4()).diagonalLength() > (first.p3() - second.p4()).diagonalLength();
+
+        return isTrailingSideOverlapping && isLeadingSideOverlapping;
```
Do both conditions have to be true for `areOverlapping`?

whsieh replied:
I think this is the right logic for this particular scenario (i.e. "two adjacent selection rects are clipping into each other, in such a way that the top and bottom edges of both rects overlap").

But you raise a good point — perhaps, `areOverlapping` should renamed to something like `leadingAndTrailingEdgesOverlap`.

rr-codes replied:
yes, otherwise it may not be "properly" overlapping

aprotyas replied:
Ooh I get what's happening now. @rr-codes going to forward a minor optional suggestion from Wenson's comment - can you rename `areOverlapping` to something like `leadingAndTrailingEdgesOverlap`?

rr-codes committed:
Selection rects for Live Text for images with vertical writing mode are sometimes incorrect
https://bugs.webkit.org/show_bug.cgi?id=255625
rdar://106205724

Reviewed by Wenson Hsieh and Megan Gardner.

For Live Text, always use `UseIndivudalQuads` so that non-rectilinear selections always work.

* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::adjustLineHeightOfSelectionGeometries):
(WebCore::RenderObject::collectSelectionGeometriesInternal):
(WebCore::coalesceSelectionGeometryWithAdjacentQuadsIfPossible):
(WebCore::RenderObject::collectSelectionGeometries):

Ensure that adjacent rects are coalesced if they overlap.

* Source/WebKit/UIProcess/ios/WKTextSelectionRect.mm:
(-[WKTextSelectionRectCustomHandleInfo initWithFloatQuad:isHorizontal:]):
(-[WKTextSelectionRectCustomHandleInfo bottomLeft]):
(-[WKTextSelectionRectCustomHandleInfo topLeft]):
(-[WKTextSelectionRectCustomHandleInfo bottomRight]):
(-[WKTextSelectionRectCustomHandleInfo topRight]):
(-[WKTextSelectionRect _customHandleInfo]):
(-[WKTextSelectionRect isVertical]):
(-[WKTextSelectionRectCustomHandleInfo initWithFloatQuad:]): Deleted.

Work around UIKit issues with the selection handle by changing the quad points.

Canonical link: https://commits.webkit.org/263159@main
